src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/SafeDTypeToStringTests.cs

```diff
@@ -0,0 +1,517 @@
+            Assert.Equal("n", DType.Number.SafeDTypeToString());
+            Assert.Equal("s", DType.String.SafeDTypeToString());
+            Assert.Equal("h", DType.Hyperlink.SafeDTypeToString());
+            Assert.Equal("d", DType.DateTime.SafeDTypeToString());
+            Assert.Equal("c", DType.Color.SafeDTypeToString());
+            Assert.Equal("$", DType.Currency.SafeDTypeToString());
+            Assert.Equal("i", DType.Image.SafeDTypeToString());
+            Assert.Equal("p", DType.PenImage.SafeDTypeToString());
+            Assert.Equal("m", DType.Media.SafeDTypeToString());
+            Assert.Equal("g", DType.Guid.SafeDTypeToString());
+            Assert.Equal("o", DType.Blob.SafeDTypeToString());
+            Assert.Equal("r*", AttachmentTableType.SafeDTypeToString());
+            Assert.Equal("r!", AttachmentRecordType.SafeDTypeToString());
+            Assert.Equal("T", DType.Time.SafeDTypeToString());
+            Assert.Equal("D", DType.Date.SafeDTypeToString());
+            Assert.Equal("N", DType.ObjNull.SafeDTypeToString());
+            Assert.Equal("P", DType.Polymorphic.SafeDTypeToString());
+            Assert.Equal("V", DType.NamedValue.SafeDTypeToString());
+            Assert.Equal("X", DType.Deferred.SafeDTypeToString());
+            Assert.Equal("-", DType.Void.SafeDTypeToString());
+            Assert.Equal("w", DType.Decimal.SafeDTypeToString());
+        }
+
+        [Fact]
+        public void RecordAndTableDTypeTests()
+        {
+            bool err = false;
+            var type1 = DType.CreateTable(
+                new TypedName(DType.Number, new DName("A")),
+                new TypedName(DType.Number, new DName("B")),
+                new TypedName(DType.Number, new DName("C")));
+
+            Assert.Equal("*[field1:n, field2:n, field3:n]", type1.SafeDTypeToString());
+            Assert.Equal("![field1:n, field2:n, field3:n]", type1.ToRecord().SafeDTypeToString());
+            Assert.Equal("![field1:n, field2:n, field3:n]", type1.ToRecord().ToRecord().SafeDTypeToString());
+            Assert.Equal("![field1:n, field2:n, field3:n]", type1.ToRecord().SafeDTypeToString());
+
+            var type2 = DType.CreateTable(
+                new List<TypedName>()
+                {
+                    new TypedName(DType.Number, new DName("B")),
+                    new TypedName(DType.Number, new DName("A")),
+                    new TypedName(DType.Number, new DName("C"))
+                });
+
+            Assert.Equal("*[field1:n, field2:n, field3:n]", type2.SafeDTypeToString());
+
+            type2 = DType.CreateRecord(type1.GetNames(DPath.Root).ToArray());
+            type2 = type2.Add(ref err, DPath.Root, new DName("B"), DType.String);
+            Assert.Equal("![field1:n, field2:s, field3:n]", type2.SafeDTypeToString());
+
+            type2 = DType.EmptyTable
+                .Add(ref err, DPath.Root, new DName("B"), DType.Number)
```

Comment:
>.Add(ref err, DPath.Root, new DName("B"), DType.Number) [](http://example.com/codeflow?start=16&length=55)

If you're not using the `err` argument, and always passing DPath.Root to the second one, you can use the smaller overload that only takes DName/DType